L58-python-async-api.md

```diff
@@ -0,0 +1,1739 @@
+            message. Request goes unserialized in case None is passed.
+          response_deserializer: Optional behaviour for deserializing the
+            response message. Response goes undeserialized in case None is
+            passed.
+
+        Returns:
+          A UnaryStreamMultiCallable value for the name server streaming method.
+        """
+
+    def stream_unary(self,
+                     method: Text,
+                     request_serializer: Optional[Callable[[Any], bytes]]=None,
+                     response_deserializer: Optional[Callable[[bytes], Any]]=None
+        ) -> grpc.aio.StreamUnaryMultiCallable:
+        """Creates a StreamUnaryMultiCallable for a client streaming method.
+
+        Args:
+          method: The name of the RPC method.
+          request_serializer: Optional behaviour for serializing the request
+            message. Request goes unserialized in case None is passed.
+          response_deserializer: Optional behaviour for deserializing the
+            response message. Response goes undeserialized in case None is
+            passed.
+
+        Returns:
+          A StreamUnaryMultiCallable value for the named client streaming method.
+        """
+
+    def stream_stream(self,
+                      method: Text,
+                      request_serializer: Optional[Callable[[Any], bytes]]=None,
+                      response_deserializer: Optional[Callable[[bytes], Any]]=None
+        ) -> grpc.aio.StreamStreamMultiCallable:
+        """Creates a StreamStreamMultiCallable for a bi-directional streaming method.
+
+        Args:
+          method: The name of the RPC method.
+          request_serializer: Optional behaviour for serializing the request
+            message. Request goes unserialized in case None is passed.
+          response_deserializer: Optional behaviour for deserializing the
+            response message. Response goes undeserialized in case None
+            is passed.
+
+        Returns:
+          A StreamStreamMultiCallable value for the named bi-directional streaming method.
+        """
+
+    def close(self) -> None:
+        """Closes this Channel and releases all resources held by it.
+
+        Closing the Channel will immediately terminate all RPCs active with the
+        Channel and it is not valid to invoke new RPCs with the Channel.
+
+        This method is idempotent.
+        """
+```
+
+```Python
+# grpc.aio.UnaryUnaryMultiCallable
+class UnaryUnaryMultiCallable:
+    """Affords invoking an async unary RPC from client-side."""
+
+    def __call__(self,
+                 request: Any,
+                 timeout: Optional[int]=None,
+                 metadata: Optional[Sequence[Tuple[Text, Text]]]=None,
+                 credentials: Optional[grpc.CallCredentials]=None,
+                 wait_for_ready: Optional[bool]=None,
+                 compression: Optional[grpc.Compression]=None
+    ) -> grpc.aio.Call[Any]:
```

Comment:
1. Please avoid `Any` in all type definitions.
Signature of this class could be this instead:
```python
class UnaryUnaryMultiCallable(Generic[Request, Response]):

    def __call__(self, request: Request, ...) -> grpc.aio.Call[Response]:
        pass
```

This allows to expose more runtime errors at compile time.

2. `metadata` should be `Sequence[Tuple[str, Union[bytes, str]]]` as grpc allows binary metadata (if suffixed by `-bin`)
3. `Text` is an alias for `str` in python3-only code, it's only useful for migrations from py2 to py3.

Reply:
Good point! Changed to `typing.AnyStr`.

Reply:
I'm not sure `typing.AnyStr` is correct, e.g. this input is valid but will be rejected by typing system:
```python
metadata=[
  ("foo", "text"),
  ("boo-bin", b"binary"),
],
```

Reply:
Can I ask for what went wrong about the `typing.AnyStr`? I wrote a snippet, and `pytype` is happy with it.

```Python
from typing import Text, AnyStr, Sequence, Tuple

metadata=[
  ("foo", "text"),
  ("boo-bin", b"binary"),
]

def process(metadata: Sequence[Tuple[Text, AnyStr]]):
    for key, value in metadata:
        print(f"I'm okay with ({type(key)} {type(value)})")

process(metadata)
```